### PR TITLE
Task3a, View All Products

### PR DESCRIPTION
**Pull request by turnerror**

### CODE REVIEW

File: app/dependencies.php [hunk 1]
```diff
@@ -36,8 +36,8 @@
     $container['Database'] = new App\Utilities\Database();
 
     $container['ProductModel'] = DI\factory('App\Factories\ProductModelFactory');
-
     $container['AddProductController'] = DI\factory('App\Factories\AddProductControllerFactory');
+    $container['GetProductsController'] = DI\factory('App\Factories\GetProductsControllerFactory');
 
     $containerBuilder->addDefinitions($container);
 };
```
Comment (alexreed22496) [inline app/dependencies.php:43]: Whitespace needed on the bottom of all files

File: src/Models/ProductModel.php [hunk 2]
```diff
@@ -14,8 +13,58 @@ class ProductModel implements ProductModelInterface
...
         $this->db = $db;
     }
+
+    /**
+     * Gets all products from Database
+     * @return array|false depending on successful query or not
+     */
+    public function getAllProducts()
+    {
+        $query = $this->db->query('SELECT `sku`, `name`, `price`, `stockLevel` FROM `products`;');
```
Comment (amandahrdmn) [inline src/Models/ProductModel.php:27]: need to add a where clause to only select products that haven't been deleted

File: src/Factories/GetProductsControllerFactory.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace App\Factories;
+
+use App\Controllers\GetProductsController;
+use Psr\Container\ContainerInterface;
+
+class GetProductsControllerFactory
+{
```
Comment (alexreed22496) [inline src/Factories/GetProductsControllerFactory.php:9]: docblock?

File: src/Controllers/GetProductsController.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    public function __construct(ProductModelInterface $productModel)
+    {
+        $this->productModel = $productModel;
+    }
+
+    public function __invoke(Request $request, Response $response, array $args)
+    {
+        $products = $this->productModel->getAllProducts();
+
+        if ($products) {
```
Comment (amandahrdmn) [inline src/Controllers/GetProductsController.php:27]: we need separate responses depending on whether we couldn't connect to the database or whether there weren't any products in it

EDIT: maybe I'm misunderstanding the query return? based on my understanding:
it's false if the query doesn't work but returns the products as an array of Product Entities if the query does work
if the query works but there aren't any products in the DB, it returns an empty array, which is evaluated as false

File: src/Factories/ProductModelFactory.php [hunk 5]
```diff
@@ -11,7 +11,7 @@ class ProductModelFactory
 {
```
Comment (alexreed22496) [inline src/Factories/ProductModelFactory.php:11]: docblock here

File: src/Models/ProductModel.php [hunk 6]
```diff
@@ -14,8 +13,58 @@ class ProductModel implements ProductModelInterface
...
+    public function __construct(\PDO $db)
     {
         $this->db = $db;
     }
+
+    /**
+     * Gets all products from Database
+     * @return array|false depending on successful query or not
+     */
+    public function getAllProducts()
```
Comment (alexreed22496) [inline src/Models/ProductModel.php:25]: if it returns as an array does this need an ": array" at the end?

File: src/Models/ProductModel.php [hunk 7]
```diff
@@ -14,8 +13,58 @@ class ProductModel implements ProductModelInterface
...
+    {
+        $query = $this->db->query('SELECT `sku`, `name`, `price`, `stockLevel` FROM `products`;');
+
+        if ($query){
+            return $query->fetchAll();
+        }
+
+        return $query;
+        }
```
Comment (amandahrdmn) [inline src/Models/ProductModel.php:34]: can maybe simplify to a ternary?

File: src/Models/ProductModel.php [hunk 8]
```diff
@@ -14,8 +13,58 @@ class ProductModel implements ProductModelInterface
...
+
+        return $query->execute($array);
+    }
+
+    /**
+     * @param string $sku
+     * @return bool whether the product was found in DB or not
+     */
+    public function checkProductExists(string $sku): bool
+    {
```
Comment (MaxGranger97) [inline src/Models/ProductModel.php:64]: This function was changed in story2 so maybe pull in the updated version before merging this

File: app/dependencies.php (hunk 1 above)
Comment (turnerror) [reply app/dependencies.php:43]: Yeah there is a whitespace at the bottom of the file, it's just not shown on github. there woul be a red error circle thingy if there wasn't

File: src/Models/ProductModel.php (hunk 6 above)
Comment (turnerror) [reply src/Models/ProductModel.php:25]: it returns an array if the query worked but false if it didn't

**Pull request merged**